src/messenger/messenger.cpp

```diff
@@ -418,25 +389,29 @@ void Messenger::RunTask() {
 }
 
 void Messenger::Terminate() {
+  if (is_messenger_running_) {
+    // Shut down the ZeroMQ context. This causes all existing sockets to abort with ETERM.
+    zmq_ctx_->shutdown();
+    MessengerUtil::DeleteLocalIPC(port_);
+    MESSENGER_LOG_INFO(fmt::format("[PID={}] Messenger terminated.", ::getpid()));
+  }
   is_messenger_running_ = false;
-  // Shut down the ZeroMQ context. This causes all existing sockets to abort with ETERM.
-  zmq_ctx_->shutdown();
-  MESSENGER_LOG_INFO(fmt::format("[PID={}] Messenger terminated.", ::getpid()));
 }
 
 void Messenger::ListenForConnection(const ConnectionDestination &target, const std::string &identity,
                                     CallbackFn callback) {
-  std::lock_guard lock(routers_add_mutex_);
+  std::unique_lock lock(routers_add_mutex_);
   // TODO(WAN): all this copying is stupid.
```

Comment:
These sorts of TODOs aren't terribly helpful 6 months in the future when someone else (or even you) is reading it. Can you give a bit more on what you'd do to fix it (if you know) and what the restriction on implementing it now is (it's okay if it's just time).

Reply:
Oh, just saw this. I think this was an old comment that wanted to avoid the indirection of 

1. T1: invokes `ListenForConnection` trying to add a new connection endpoint, which enqueues the router to be added
2. dedicated messenger thread T2: grabs it off the queue, adds it

In addressing Joe's comment, I added documentation that notes why this is necessary (iirc zeromq requires that sockets are used on the thread where they are created)